PageRepository 생성

tableMinPark opened the pull request.

tableMinPark committed:
feat : H2 초기 설정 및 초기 데이터 저장 sql 설정

tableMinPark committed:
feat : PageRepository 구현

tableMinPark committed:
feat : PageRepository 조회 단위 테스트

daengdaengLee commented on src/main/java/com/hello/page/repository/PageRepository.java:
```diff
@@ -0,0 +1,92 @@
+
+    private final NamedParameterJdbcTemplate template;
+
+    @Autowired
+    public PageRepository(NamedParameterJdbcTemplate template) {
+        this.template = template;
+    }
```
여기는 @RequiredArgsConstructor 이걸로 대체할 수 있을 것 같아요!

daengdaengLee commented on src/main/java/com/hello/page/repository/PageRepository.java:
```diff
@@ -0,0 +1,92 @@
+@Repository
+public class PageRepository {
+    private static final String SELECT_PAGE_BASE_SQL = "SELECT id, title, content, parent_id FROM page";
+    private static final String SELECT_PAGE_BY_ID_SQL = SELECT_PAGE_BASE_SQL + " WHERE id = ?";
```
여기도 name param 쓰는 게 어떨까요? 
? 대신 :id 로요. :)

daengdaengLee commented on src/main/java/com/hello/page/repository/PageRepository.java:
```diff
@@ -0,0 +1,92 @@
+    private static final String SELECT_PAGE_BY_PARENT_ID_SQL = SELECT_PAGE_BASE_SQL + " WHERE parent_id = :parentId";
+    private static final String SELECT_PAGE_WHERE_ID_IN_SQL = SELECT_PAGE_BASE_SQL + " WHERE id IN (:ids)";
+    private static final String SELECT_ANCESTORS_RECURSIVELY = """
+            WITH CHILD(ID, PARENT_ID) AS (
+              SELECT ID, PARENT_ID
+              FROM PAGE
+              WHERE ID = :id
+              UNION ALL
+              SELECT P.ID, P.PARENT_ID
+              FROM CHILD C
+              INNER JOIN PAGE P ON C.PARENT_ID = P.ID
+            )
+            SELECT *
+            FROM CHILD
+            OFFSET 1 ROWS;
```
테이블이랑 컬럼명을 소문자로 써서, 해당 키워드만 소문자로 바꾸면 덜 헷갈릴 것 같습니다! :)

CHILD -> child, C -> c
PAGE -> page, P -> p
ID -> id
PARENT_ID -> parent_id

daengdaengLee replied:
WITH 다음에 RECURSIVE 키워드 빠져도 괜찮은거죠?

tableMinPark replied:
> 테이블이랑 컬럼명을 소문자로 써서, 해당 키워드만 소문자로 바꾸면 덜 헷갈릴 것 같습니다! :) 

자꾸 오류나서 제 코드부분 넣고 그대로 올렸네요. 수정했습니다! :)

> WITH 다음에 RECURSIVE 키워드 빠져도 괜찮은거죠?

RECURSIVE 없어도 괜찮긴한데 명시적으로 붙히는게 좋을 것 같아서 수정했습니다! :)

daengdaengLee commented on src/main/java/com/hello/page/repository/PageRepository.java:
```diff
@@ -0,0 +1,92 @@
+        List<PageInfo> pageInfos = template.query(
+                SELECT_ANCESTORS_RECURSIVELY,
+                param,
+                (rs, rowNum) ->  PageInfo.builder()
+                    .id(rs.getLong("id"))
+                    .parentId(rs.getLong("parent_id"))
+                    .build());
```
여기서 아마 rs.wasNull 을 써서 좀 지저분하게 처리해야 헀던 것 같아요.
안 그러면 마지막 루트 페이지에서 parentId 가 0 으로 나오더라구요.

daengdaengLee replied:
추가로 이 부분은 PageInfo 를 쓰기보단 Repository 내부에서 static inner record 를 만들어서 쓰는 게 좋을 것 같습니다!

tableMinPark replied:
> 여기서 아마 rs.wasNull 을 써서 좀 지저분하게 처리해야 헀던 것 같아요.

auto_Increment 라 0은 기존 설정으로는 나올 수 없어서 없애도 될 것 같았는데 null로 바꾸는게 더 나을 것 같네요! 
수정하겠습니다!

tableMinPark replied:
> 추가로 이 부분은 PageInfo 를 쓰기보단 Repository 내부에서 static inner record 를 만들어서 쓰는 게 좋을 것 같습니다!

record 로 만들어서 쓰는 방식으로 수정하겠습니다! :)

daengdaengLee replied:
아 요게 이유가, DB 에 parent_id 는 null 인 부분이 (루트 페이지 같은 경우) ResultSet 에서 getLong 으로 읽어버리면 기본값인 0L 로 튀어나와서 그런 것 같아요!

daengdaengLee commented on src/main/java/com/hello/page/repository/PageRepository.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    private PageInfo pageRowMapper(ResultSet rs, int rowNum) throws SQLException {
+        Long id = rs.getLong("id");
+        String title = rs.getString("title");
+        String content = rs.getString("content");
+        Long parentId = rs.getLong("parent_id");
```
여기도 위에 코멘트처럼 rs.wasNull 로 처리해줘야 할 것 같아요!

tableMinPark replied:
> 여기도 위에 코멘트처럼 rs.wasNull 로 처리해줘야 할 것 같아요!

동일하게 수정하겠습니다!

daengdaengLee commented on src/main/resources/data.sql:
요 부분은 제가 작성했던 것처럼 1자 모양보단 트리 모양이 나오게 조금 더 신경쓰는 게 좋을 것 같아요.
SQL 로 처리하면 생성된 id 매핑하기가 어려워서 저는 자바 코드로 했었어요.

daengdaengLee commented on src/test/java/com/hello/page/repository/PageRepositoryTest.java:
```diff
@@ -0,0 +1,48 @@
+    @Test
+    @DisplayName("id를 기준으로 페이지를 조회한다.")
+    void findByIdTest() {
+        PageInfo pageInfo = pageRepository.findById(1L).get();
+        assertNotNull(pageInfo);
```
바로 get 한 후에 검증하는 것보단, Optional 로 받아서 isPresent() 가 true 인지 검사하는 게 좋을 것 같습니다.

daengdaengLee replied:
빈 Optional 인데 바로 get 하면 에러 터져서요. :)

daengdaengLee commented on src/test/java/com/hello/page/repository/PageRepositoryTest.java:
```diff
@@ -0,0 +1,48 @@
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
```
이거 AssertJ 쓰는 게 어떨까요?
거의 다 AssertJ 쓰는 것 같더라구요

daengdaengLee commented on src/test/java/com/hello/page/repository/PageRepositoryTest.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @Test
+    @DisplayName("id 목록을 기준으로 페이지 목록을 조회한다.")
```
설명이 약간 내용이랑 안 맞는 것 같습니다.

daengdaengLee commented on src/test/java/com/hello/page/repository/PageRepositoryTest.java:
```diff
@@ -0,0 +1,48 @@
+    @Test
+    @DisplayName("pageId 를 기준으로 root 페이지까지의 id를 순서대로 조회한다.")
+    void findParentPageIdsTest() {
+        List<Long> ids = pageRepository.findParentPageIds(10L);
+        assertEquals(9, ids.size());
```
여기가 아마 위에 Repository 에서 wasNull 처리해서 거르면,
루트 페이지는 부모 아이디가 없어서 길이 8로 나올 것 같아요.

daengdaengLee reviewed (approved):
LGTM!
코멘트 확인 부탁드릴게요!

tableMinPark committed:
feat : 데이터 생성을 위한 Runner 생성

tableMinPark committed:
feat : PageRepository 수정

- pageRowMapper rs.wasNull 처리 (값이 없으면 null로 변환)
- root 페이지까지의 id 목록 조회할 때 자기 자신을 OFFSET 으로 제외시키고, 남은 목록의 id를 조회하는 방식으로 변경

tableMinPark committed:
feat : PageRepositoryTest 수정

- AssertJ 로 변경
- 데이터 생성 Runner 적용 후 단위 테스트

daengdaengLee commented on src/main/java/com/hello/page/repository/InitDataRunner.java:
```diff
@@ -0,0 +1,75 @@
+    private final NamedParameterJdbcTemplate template;
+    private final KeyHolder keyHolder = new GeneratedKeyHolder();
+
+    private final int final_level = 5;
```
여기 카멜케이스로 바꾸는 게 좋을 것 같습니다!

daengdaengLee commented on src/main/java/com/hello/page/repository/PageRepository.java:
```diff
@@ -0,0 +1,95 @@
+        return template.query(SELECT_PAGE_BY_PARENT_ID_SQL, param, this::pageRowMapper);
+    }
+
+    public List<PageInfo> findAllById(Iterable<Long> ids) {
+        MapSqlParameterSource param = new MapSqlParameterSource().addValue("ids", ids);
+        return template.query(SELECT_PAGE_WHERE_ID_IN_SQL, param, this::pageRowMapper);
+    }
```
ids 길이 먼저 검사해서 길이가 0 이면 (비어있으면)
쿼리 날리지 말고 바로 빈 리스트 반환하면 어떨까요?

daengdaengLee commented on src/main/java/com/hello/page/repository/PageRepository.java:
```diff
@@ -0,0 +1,95 @@
+
+    private PageInfo pageRowMapper(ResultSet rs, int rowNum) throws SQLException {
+        Long id = rs.getLong("id");
+        if (rs.wasNull()) {
+            id = null;
+        }
```
이건 빼도 될 것 같습니다.
나머지 컬럼은 nullable 이지만 id 는 pk 라 상민님께서 앞전에 코멘트 주신대로 저거 체크할 필요는 없을 것 같아요!

daengdaengLee commented on src/test/java/com/hello/page/repository/PageRepositoryTest.java:
```diff
@@ -0,0 +1,51 @@
+    @DisplayName("id를 기준으로 페이지를 조회한다.")
+    void findByIdTest() {
+        Optional<PageInfo> pageInfoOptional = pageRepository.findById(1L);
+        assertThat(pageInfoOptional.isPresent()).isFalse();
```
오잉? 이거 isTrue 로 검사하는 게 맞지 않을까요?

daengdaengLee commented on src/test/java/com/hello/page/repository/PageRepositoryTest.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Test
+    @DisplayName("id 목록을 기준으로 서브 페이지 목록을 조회한다.")
```
"id 를 기준으로 서브 페이지 목록을 조회한다." 가 더 맞을 것 같아요.

daengdaengLee reviewed (approved):
LGTM!
코멘트 몇개만 확인해주세요! 고생하셨어요!

tableMinPark committed:
feat : initDataRunner 변수명 수정

- 카멜케이스로 수정

tableMinPark committed:
feat : 상위 페이지가 없는 경우 쿼리 날리지 않도록 수정

- Iterator에서 요소를 가진 경우에만 쿼리를 날리도록 수정
- 빈 리스트를 선언하여 요소가 없는 경우 빈 리스트 반환

tableMinPark committed:
feat : 단위 테스트 검증 오류 수정

- findByIdTest 에서 null인 경우 false -> true로 수정
- findByParentIdTest DisplayName 수정
